src/_P098_CacheReader.ino

```diff
@@ -0,0 +1,478 @@
+      this->fetchSampleIndex();
+    }
+    if (*offset < 1){
+      this->getOffset();
+    }
+    this->setData();
+  }
+  void fetchFileNr(){
+    // TODO: fetch file nr from server
+    (*this->fileNr) = 1;
+  }
+  void fetchSampleIndex(){
+    // TODO: fetch current sample index from server
+  }
+  void getOffset(){
+    // TODO: Find offset instead of static 16 byte offset
+    float staticOffset = 16;
+    memcpy(offset, &staticOffset, 4);
+  }
+  bool setData(){
+    /*
+    char *filename = new char[24];
+    for (int i = 0 ; i<24; i++){
+      filename[i] = '\0';
+    }
+    sprintf(filename,"cache_%d.bin",fnr);
+    */
+    int fnr = (int)(*this->fileNr);
+    char *filename = returnFilename(fnr);
+
+
+    if (!fileExists(filename)){
+      this->nextFile();
+      return false;
+    }
+    cache = tryOpenFile(filename,"r");
+    fileOpen = true;
+
+    delete[] filename;
+    return true;
+  }
+  void readSample(struct EventStruct *event){
+    if (fileOpen){
+    // Seek to current sample index
+    // TODO: Check return value if seek is out of bounds!
+    cache.seek(24 * (*sampleIndex) + (*offset));
+    // Read sample data into buffer
+    size_t read = cache.read(buffer, 24);
+    // Check if read was valid and does not contain EOF
+    if (read == 24 && buffer[23] != EOF){
+      // Parse buffer into Sample_t
+      sample->parseSample(buffer);
+      // Validate and send sample
+      bool validated = sample->validateSample();
+      if (validated){
+        sample->sendSample(event);
+      }
+      // Increment to next sample
+      this->increment();
+    } else {
+      this->deleteFile();
+      this->nextFile();
+    }
+    }else {
+      this->nextFile();
+    }
+  }
+  void setIndex(int index){
+    // TODO:
+  }
+  void increment(){
+    // TODO: Check if last sample in file
+    // If so, move to next file
+    (*this->sampleIndex)++;
+  }
+  void deleteFile(){
+    /*
+    int fnr = (int)(*this->fileNr);
+    char *filename = new char[24];
+    for (int i = 0 ; i<24; i++){
+      filename[i] = '\0';
+    }
+    sprintf(filename,"cache_%d.bin",fnr);
+    */
+    int fnr = (int)(*this->fileNr);
+    char * filename = returnFilename(fnr);
+
+    tryDeleteFile(filename);
+    delete[] filename;
+  }
+  char *returnFilename(int filenr){
+    char *filename = new char[24];
+    for (int i = 0 ; i<24; i++){
+      filename[i] = '\0';
+    }
+    sprintf(filename,"cache_%d.bin",filenr);
+    return filename;
+  }
+  void nextFile(){
+    if (fileOpen){
+      cache.close();
+      fileOpen = false;
+    }
+    (*this->fileNr)++;
+    int fnr = (int)(*this->fileNr);
+    char *fname = new char[24];
+    for (int i = 0 ; i<24; i++){
+      fname[i] = '\0';
+    }
+    sprintf(fname,"cache_%d.bin",fnr);
+    if (fileExists(fname)){
+      (*this->sampleIndex) = 0;
+      (*this->fileNr) = fnr;
+      //this->setData();
+      //cache.close();
+      cache = tryOpenFile(fname,"r");
+      fileOpen = true;
```

Comment:
`fileOpen` does not reflect the true state.
Maybe you can remove the fileOpen member as you also keep a file handler.
This handler can check whether it is "open" or not.